### PR TITLE
FrxEth Redeem Strategy

### PR DESCRIPTION
**Pull request by DanielVF**
This will allow us to convert amounts of FrxEth to WETH via the Frax ETH redemption queue.

### CODE REVIEW

**Comment by DanielVF** [comment]
```python
from world import *
FRXETH_REDEEM_STRAT = '0xB262b69d1dB3dc092C73384B8553DA3d00e93682'
frxeth_redeem_strat = load_contract('frxeth_redeem_strat', FRXETH_REDEEM_STRAT)
oeth_vault_admin.withdrawAllFromStrategy(FRAX_ETH_STRATEGY, {'from': STRATEGIST})
print(frxeth.balanceOf(vault_oeth_core) / 1e18, weth.balanceOf(vault_oeth_core) / 1e18)
vault_oeth_admin.depositToStrategy(FRXETH_REDEEM_STRAT, [frxeth], [1003e18], {'from': STRATEGIST})
tickets = [int(x['data'][2:66],16) for x in history[-1].logs if x['address'] == FRXETH_REDEEM_STRAT and len(x['topics'])==1]
print(tickets)
chain.sleep(16*24*3600)
frxeth_redeem_strat.redeemTickets(tickets, 1003e18, {'from': STRATEGIST})
```

**Review by naddison36 — APPROVED** [review]
This looks good.

I've pushed some changes to fix the prettier, linter, slither and the unit tests.
The fork tests are currently failing but that's to do with OUSD's Flux strategy.

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+/**
+ * @title Frax ETH Redeem Strategy
+ * @notice This strategy redeems Frax ETH for ETH via the Frax Eth Redemption Queue contract
+ */
+contract FrxEthRedeemStrategy is InitializableAbstractStrategy {
+    IWETH9 private constant weth =
+        IWETH9(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+    IERC20 private constant frxETH =
+        IERC20(0x5E8422345238F34275888049021821E8E08CAa1f);
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:31]: Yes, for most things I would use immutable as well, but for this it is purely going to be fork tested, and makes no sense out of a mainnet context.

**Comment by DanielVF** [comment]
Thanks for the cleanup, Nick.

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,233 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC721Receiver } from "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+
+import { BaseCompoundStrategy, InitializableAbstractStrategy } from "./BaseCompoundStrategy.sol";
```
Comment (sparrowDom) [inline contracts/contracts/strategies/FrxEthRedeemStrategy.sol:7]: the `BaseCompoundStrategy` seems not to be used

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+import { IComptroller } from "../interfaces/IComptroller.sol";
```
Comment (sparrowDom) [inline contracts/contracts/strategies/FrxEthRedeemStrategy.sol:8]: also an unused import

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,233 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
```
Comment (sparrowDom) [inline contracts/contracts/strategies/FrxEthRedeemStrategy.sol:4]: unused import

**Review by sparrowDom — DISMISSED** [review]
good job lgtm

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol (hunk 2 above)
Comment (naddison36) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:7]: I'll fix these are I've got the branch checked out

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol (hunk 4 above)
Comment (naddison36) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol]: fixed

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol (hunk 3 above)
Comment (naddison36) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol]: fixed

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        // slither-disable-next-line unused-return
+        frxETH.approve(address(redemptionQueue), type(uint256).max);
+    }
+
+     /**
+     * @notice Check if an asset is supported.
+     * @param _asset    Address of the asset
+     * @return bool     Whether asset is supported
+     */
+    function supportsAsset(address _asset) public view override returns (bool) {
```
Comment (shahthepro) [inline contracts/contracts/strategies/FrxEthRedeemStrategy.sol:222]: Visibility can be set to `pure`

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    /**
+     * @notice Returns the amount of queued FraxEth that will be returned as WETH.
+     * We return this as a WETH asset, since that is what it will eventually be returned as.
+     * We only return the outstandingRedeems, because the contract itself should never hold any funds.
+     * @param _asset      Address of the asset
+     * @return balance    Total value of the asset in the platform
+     */
+    function checkBalance(address _asset)
```
Comment (shahthepro) [inline contracts/contracts/strategies/FrxEthRedeemStrategy.sol:195]: nit: For all other strategies, `checkBalance` reverts if asset isn't supported. This is a temporary strategy, so it shouldn't matter though

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        uint256 currentBalance = payable(address(this)).balance;
+        uint256 redeemedAmount = currentBalance - startingBalance;
+        require(
+            expectedAmount == redeemedAmount,
+            "Redeemed amount does not match expected amount"
+        );
+        outstandingRedeems -= redeemedAmount;
+        weth.deposit{ value: currentBalance }();
+        // slither-disable-next-line unchecked-transfer
+        weth.transfer(vaultAddress, currentBalance);
```
Comment (shahthepro) [inline contracts/contracts/strategies/FrxEthRedeemStrategy.sol:160]: Can we emit a `Withdraw` event here? Would be easier to track it with analytics that way

Comment (DanielVF) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:160]: Added

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol (hunk 6 above)
Comment (DanielVF) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:202]: Changed to match

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol (hunk 5 above)
Comment (DanielVF) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol]: Done

**Review by shahthepro — COMMENTED** [review]
A couple more comments related to gas savings, but nothing major

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+     * @notice Redeem specific tickets from the Queue.
+     * Called by the strategist.
+     * @param _nftIds Array of NFT IDs to redeem
+     */
+    function redeemTickets(uint256[] memory _nftIds, uint256 expectedAmount)
+        external
+        nonReentrant
+    {
+        require(
+            msg.sender == IVault(vaultAddress).strategistAddr(),
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:138]: I'd rather have it just update and always follow the latest value on the vault. Plus it's only used in an admin action, so we don't have to worry about gas.

For our non-default strategies, the only gas important methods are supportsAsset and checkBalance.

**Review by shahthepro — DISMISSED** [review]
#### Requirements
- Adds a one-off strategy to redeem frxETH into ETH using Frax's redemption queue

### Deployment Considerations

#### Internal State

- What can be always said about relationships between stored state
- What must hold true about state before a function can run correctly (preconditions)
- What must hold true about the return or any changes to state after a function has run.

Does this code do that?

#### Attack

- ⚪ An malicious actor could send their redeemable NFT to the strategy. And if the strategist accidentally uses that NFT while calling `redeemTickets` method, it could cause underflow of `outstandingRedeems`. However this will not result in loss of funds in any way. Worst case, we will have to upgrade the contract in future. Could be totally avoided if strategist takes necessary precaution to only use NFTs that the strategy minted. **No need to fix**


#### Logic

  - [x] _Are there bugs in the logic?_: **No logical bugs**
  - [x] _Correct usage of global & local variables. -> they might differentiate only by an underscore that can be overlooked (e.g. address vs _address)._: The contract initialises some of the storage variables to `address(0)`, however it has zero impact on overall functioning of the contract and doesn't result in any bugs. So, it's fine
  - [x] _Do public/external functions having asset parameter verify the asset is supported (where applicable)_ **Yes**


#### Tests

  - [ ] Each publicly callable method has a test
  - [ ] Each logical branch has a test
  - [ ] Each require() has a test
  - [x] Edge conditions are tested
  - [x] If tests interact with AMM make sure enough edge cases (pool tilts) are tested. Ideally with fuzzing.

#### Flavor

_Could this code be simpler?_ All good

_Could this code be less vulnerable to other code behaving weirdly?_ All good

#### Overflow

- [x] Never use "+" or "-", always use safe math or have contract compile in solidity version > 0.8
- [x] Check that all for loops use uint256

#### Proxy
- [x] Make sure proxy implementation contracts don't initialize variable state on variable declaration and do it rather in initialize function.

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying
- [x] (If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing.)
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Dependencies
- [x] Review any new contract dependencies thoroughly (e.g. OpenZeppelin imports) when new dependencies are added or version of dependencies changes.
- [x] If OpenZeppelin ACL roles are use review & enumerate all of them.
- [x] Check OpenZeppelin [security vulnerabilities](https://github.com/OpenZeppelin/openzeppelin-contracts/security/advisories) and see if any apply to current PR considering the version of OpenZeppelin contract used.

#### Deploy
- [ ] Check that any deployer permissions are removed after deploy

#### Authentication

- [x] Never use tx.origin
- [ ] ⚪ _Check that every external/public function should actually be external._ **`deposit()` could be `external`. But not really a blocker**
- [x] Check that every external/public function has the correct authentication

#### Cryptographic code

- [x] Contracts that roll their own crypto are terrifying
- [x] Note that a failed signature check will result in a 0x00 result. Make sure that the result throws if it returns this.
- [x] Beware of signed data being used in a replay attack to other contracts.

#### Gas problems

- [x] Contracts with for loops must have either:
    - [x] A way to remove items
    - [x] Can be upgraded to get unstuck
    - [x] Size can only controlled by admins
- [x] Contracts with for loops must not allow end users to add unlimited items to a loop that is used by others or admins.

#### External calls

- [x] Contract addresses passed in are validated
- [x] No unsafe external calls
- [x] Reentrancy guards on all state changing functions
    - [ ] Still doesn't protect against external contracts changing the state of the world if they are called.
- [x] Malicious behaviors
- [x] Could fail from stack depth problems (low level calls must require success)
- [x] No slippage attacks (we need to validate expected tokens received)
- [x] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Don't call balanceOf for external contracts to determine what they will do, when they instead use internal accounting?


#### Ethereum

- [x] Contract wraps ETH and can also receive ETH from the redemption queue
- [x] Contract has no payable methods (other than fallback)

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (shahthepro) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:138]: Sounds good

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+        // strategist via redeemTickets
+        require(false, "use redeemTickets() instead");
+    }
+
+    /**
+     * @notice Redeem specific tickets from the Queue.
+     * Called by the strategist.
+     * @param _nftIds Array of NFT IDs to redeem
+     */
+    function redeemTickets(uint256[] memory _nftIds, uint256 expectedAmount)
```
Comment (shahthepro) [inline contracts/contracts/strategies/FrxEthRedeemStrategy.sol:133]: nit: Could probably add a `expectedAmount > 0` check

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+        // slither-disable-next-line unchecked-transfer
+        weth.transfer(vaultAddress, currentBalance);
+        emit Withdrawal(
+            address(weth),
+            address(redemptionQueue),
+            currentBalance
+        );
+    }
+
+    function _abstractSetPToken(address _asset, address _pToken)
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:168]: I've added a revert on _abstractSetPToken which will prevent a pToken set from happening. I've not adjusted remove, since if there are no adds a remove should not work.

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    function redeemTickets(uint256[] memory _nftIds, uint256 expectedAmount)
+        external
+        nonReentrant
+    {
+        require(
+            msg.sender == IVault(vaultAddress).strategistAddr(),
+            "Caller is not the Strategist"
+        );
+        uint256 startingBalance = payable(address(this)).balance;
+        for (uint256 i = 0; i < _nftIds.length; i++) {
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:142]: Not really gas important here, and future solidity compilers will optimize this to be the same thing.

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+        IWETH9(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+    IERC20 private constant frxETH =
+        IERC20(0x5E8422345238F34275888049021821E8E08CAa1f);
+    IFraxEtherRedemptionQueue private constant redemptionQueue =
+        IFraxEtherRedemptionQueue(0x82bA8da44Cd5261762e629dd5c605b17715727bd);
+    uint256 public constant maxRedeemTicket = 250e18;
+    uint256 public outstandingRedeems;
+
+    event RedeemNFTMinted(uint256 _nftId, uint256 _amount);
+    event RedeemNFTBurned(uint256 _nftId);
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:35]: For which reason? To know the state of the contract and what you could redeem you would need to fetch all RedeemNFTMinted and RedeemNFTBurned events.

File: contracts/contracts/strategies/FrxEthRedeemStrategy.sol (hunk 9 above)
Comment (DanielVF) [reply contracts/contracts/strategies/FrxEthRedeemStrategy.sol:133]: Will keep as is on this one.

**Review by shahthepro — DISMISSED** [review]

**Comment by DanielVF** [comment]
# FrxEth Redeem Code Review

#### Authentication

- [x] Never use tx.origin
- [x] Every external/public function should be intended to be externally accessible
- [x] Every external/public function has the correct authentication

🟣 `deposit` could be external, rather than public. Fine as is though.

🟣 `oneERC721Received` could be view/pure. Fine as is though.


#### Ethereum

_Contract does have to receive Ethereum, however in normal operation the contract immediately wraps it and sends it to the vault as WETH.

#### Cryptographic code

- [x] This contract code does not roll it's own crypto.
- [x] No signature checks without reverting on a 0x00 result.
- [x] No signed data could be used in a replay attack, on our contract or others.

#### Gas problems

- [x] Contracts with for loops must have either:
    - [ ] A way to remove items
    - [ ] Can be upgraded to get unstuck
    - [x] Size can only controlled by admins
- [x] Contracts with for loops must not allow end users to add unlimited items to a loop that is used by others or admins.

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying. _If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing._
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Overflow

- [x] Code is solidity version >= 0.8.0
- [x] Check that all for loops use uint256

#### Proxy

- [x] No storage variable initialized at definition when contract used as a proxy implementation.

#### Events
- [x] All state changing functions emit events

## Medium Checks

#### Rounding
- [x] Contract rounds in the protocols favor
- [x] Contract does not have bugs from loosing rounding precision
- [x] Code correctly multiplies before division
- [x] Contract does not have bugs from zero or near zero amounts

#### Dependencies
_No new dependancies_

#### External calls

- [x] Contract addresses passed in are validated
- [x] No unsafe external calls
- [x] Reentrancy guards on all state changing functions
    - [ ] Still doesn't protect against external contracts changing the state of the world if they are called.
- [ ] No Malicious behaviors
- [x] Low level call() must require success.
- [x] No slippage attacks (we need to validate expected tokens received)
- [x] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Do not call balanceOf for external contracts to determine what they will do when they use internal accounting

#### Deploy

- [ ] Deployer permissions are removed after deploy

## Strategy Specific

#### Smart contract

- [x] **Check balance cannot be manipulated up AND down by an attacker**
- [x] No funds fall into the cracks and are not reported in check balance as a result of depositing or withdrawing
- [x] All funds can be recovered from the strategy by some combination of depositAll, withdraw, or withdrawAll()
- [ ] WithdrawAll() can always withdraw an amount equal to or larger than checkBalances report, even in spite of attacker manipulation. (🟣 We inherently can't, since these are asynchronous redeems)
- [x] WithdrawAll() cannot be MEV'd

#### Downstream

_Even without a backend system, we can just look at events fired from mint transactions to know what we can redeem.

## Thinking

#### Logic

  - [x] Logic free of bugs
  - [x] Correct usage of global & local variables. -> they might differentiate only by an underscore that can be overlooked (e.g. address vs _address).
  - [x] Do public/external functions having asset parameter verify the asset is supported (where applicable)

#### Internal State

`outstandingRedeems` must match the sum of all minted NFTs minus the sum of redeemed NFTs. It appears to do so.

#### Attack

Possible Impact. This contract may be moving and accounting for a large portion of OETH's funds. Permanently losing funds would be bad.

Attacker theft - The contract should hold no funds at rest between strategist operations. This limits theft operations to inside deposit and withdraw operations. Both of these can only be initiated by the strategist or governance. All funds are transferred to only the vault or allowed to be transferred to the frxeth redeem queue.

Attacker balance manipulation - Donations to the strategy do not alter the reported funds. NFT donations _that are used by the strategist_ would result in a profit to the protocol, though would results in other NFT's not being able to be redeeming until accounting could be fixed with a contract upgrade.

Given that strategy funds are empty outside of actions, and actions can only kicked off by trusted users, the most interesting attack angles would be either directly outside the actions (sandwiches, frontrunning, backrunning) or inside the action itself. 

It's possible for someone to redeem just in front of us in the frxETH pool, and thus bring the redeemable total down under what we are trying to redeem in a transaction. But we can just resubmit with a private transaction.

How about inside attacks? redeemTickets does an adjacent decrement of outstandingRedeems with sending funds to the vault. There shouldn't be any time when the vault balances are incorrect from an external point of view.

depositAll does get frxEth sent to it before the deposit happens and before outstandingRedeems is updated. This would result from the outside like a temporary drop in vault total supply. However there is no way for an attacker to get execution during this time, the vault is reentrancy locked during depositToStrategy, and a temporary down in funds should not be exploitable like a temporary up in funds would be.

NFT's provide a reentrancy danger, in that transferring them can give execution to the destination address. However we never transfer the NFTs, so this is not an issue.

Deposit process failure - It is possible to send too much funds to the deposit function that you could run out of gas creating NFT's. If this was a strategist action, then the revert would put the funds back in the vault and a smaller amount could be chosen. If someone donated a huge amount of frxETH to the strategy to cause this, we could withdraw the frxETH to vault via withdrawAll() and be very happy.

Redeem process failure - Too many NFT's could cause out of gas, but strategist could just adjust number of NFTs in redeemTickets call.

#### Flavor

There was a bit of tradeoff between writing just enough code to do the job, and handling cases like unexpected donations of eth, or incorrect governance behavior. Code seems reasonable.

**Comment by DanielVF** [comment]
I had an unsigned commit, and needed to force push to remove it as a merge blocker. Also added a lint fix.

Needless to say, re-review the last set of commits carefully before merging.

**Review by shahthepro — APPROVED** [review]
